Mathlib/RingTheory/Perfection.lean

```diff
@@ -28,140 +28,216 @@ universe u₁ u₂ u₃ u₄
-  (Ring.perfectionSubring R p).toCommRing
+instance instCommMonoid (M : Type*) [CommMonoid M] (p : ℕ) : CommMonoid (Perfection M p) :=
+  (submonoid M p).toCommMonoid
 
-instance : Inhabited (Ring.Perfection R p) := ⟨0⟩
+variable (M : Type*) [CommMonoid M] (p : ℕ)
 
 /-- The `n`-th coefficient of an element of the perfection. -/
-def coeff (n : ℕ) : Ring.Perfection R p →+* R where
+def coeffMonoidHom (n : ℕ) : Perfection M p →* M where
   toFun f := f.1 n
   map_one' := rfl
   map_mul' _ _ := rfl
-  map_zero' := rfl
-  map_add' _ _ := rfl
 
-variable {R p}
+/-- The `p`-th root of an element of the perfection. -/
+def pthRootMonoidHom : Perfection M p →* Perfection M p where
+  toFun f := ⟨fun n => coeffMonoidHom M p (n + 1) f, fun _ => f.2 _⟩
+  map_one' := rfl
+  map_mul' _ _ := rfl
+
+variable {M p}
 
 @[ext]
-theorem ext {f g : Ring.Perfection R p} (h : ∀ n, coeff R p n f = coeff R p n g) : f = g :=
+theorem ext' {f g : Perfection M p} (h : ∀ n, coeffMonoidHom M p n f = coeffMonoidHom M p n g) :
+    f = g :=
   Subtype.eq <| funext h
 
-variable (R p)
+@[simp]
+theorem coeffMonoidHom_mk (f : ℕ → M) (hf) (n : ℕ) : coeffMonoidHom M p n ⟨f, hf⟩ = f n := rfl
+
+theorem coeffMonoidHom_pthRootMonoidHom (f : Perfection M p) (n : ℕ) :
+    coeffMonoidHom M p n (pthRootMonoidHom M p f) = coeffMonoidHom M p (n + 1) f := rfl
+
+theorem coeffMonoidHom_pow_p (f : Perfection M p) (n : ℕ) :
+    coeffMonoidHom M p (n + 1) (f ^ p) = coeffMonoidHom M p n f := by
+  rw [map_pow]; exact f.2 n
+
+theorem coeffMonoidHom_pow_p' (f : Perfection M p) (n : ℕ) :
+    coeffMonoidHom M p (n + 1) f ^ p = coeffMonoidHom M p n f :=
+  f.2 n
+
+theorem coeffMonoidHom_pow_p_pow (f : Perfection M p) (m n : ℕ) :
+    coeffMonoidHom M p (m + n) (f ^ p ^ n) = coeffMonoidHom M p m f :=
+  n.recOn (by simp) fun n ih ↦ by rw [pow_succ, pow_mul, Nat.add_succ, coeffMonoidHom_pow_p, ih]
+
+theorem coeffMonoidHom_powMonoidHom (f : Perfection M p) (n : ℕ) :
+    coeffMonoidHom M p (n + 1) (powMonoidHom p f) = coeffMonoidHom M p n f :=
+  coeffMonoidHom_pow_p f n
+
+theorem coeffMonoidHom_iterate_powMonoidHom (f : Perfection M p) (n m : ℕ) :
+    coeffMonoidHom M p (n + m) ((powMonoidHom p)^[m] f) = coeffMonoidHom M p n f :=
+  m.recOn rfl fun m ih ↦ by
+    rw [Function.iterate_succ_apply', Nat.add_succ, coeffMonoidHom_powMonoidHom, ih]
+
+theorem coeffMonoidHom_iterate_powMonoidHom' (f : Perfection M p) (n m : ℕ) (hmn : m ≤ n) :
+    coeffMonoidHom M p n ((powMonoidHom p)^[m] f) = coeffMonoidHom M p (n - m) f := by
+  rw [← coeffMonoidHom_iterate_powMonoidHom f (n - m) m, Nat.sub_add_cancel hmn]
+
+theorem pthRootMonoidHom_powMonoidHom : (pthRootMonoidHom M p).comp (powMonoidHom p) = .id _ := by
+  ext; simp [coeffMonoidHom_pthRootMonoidHom, coeffMonoidHom_pow_p']
+
+theorem powMonoidHom_pthRootMonoidHom : (powMonoidHom p).comp (pthRootMonoidHom M p) = .id _ :=
+  pthRootMonoidHom_powMonoidHom
```

Comment:
Can you give these as `LeftInverse` and `RightInverse` (unbundled)? And then you'd use that in the `bijective` proof below?